lib/classes/ConfigSchemaHandler/normalizeAjvErrors.test.js

```diff
@@ -0,0 +1,563 @@
+'use strict';
+const Ajv = require('ajv');
+
+const expect = require('chai').expect;
+const normalizeAjvErrors = require('./normalizeAjvErrors');
+
+describe('#normalizeAjvErrors', () => {
+  const ajv = new Ajv({ allErrors: true });
+  let schema;
+  let userConfig;
+  let ajvErrors;
+
+  it('should show error messages for root and second level property', () => {
+    userConfig = {
+      foo: 'bar',
+      package: { incclude: ['./folder'] },
+    };
+    schema = {
+      type: 'object',
+      properties: {
+        package: {
+          type: 'object',
+          properties: {
+            include: { type: 'array', items: { type: 'string' } },
+          },
+          additionalProperties: false,
+        },
+      },
+      additionalProperties: false,
+    };
+    const validate = ajv.compile(schema);
+    validate(userConfig);
+    ajvErrors = validate.errors;
+    expect(normalizeAjvErrors(ajvErrors, userConfig).map(err => err.message)).to.deep.equal([
+      "Unrecognized property 'foo' on 'root'",
+      "Unrecognized property 'incclude' on 'package'",
+    ]);
+  });
+
+  it('should show error messages for function events related config ', () => {
+    userConfig = {
+      functions: {
+        someFunc: {
+          events: [
+            { httpp: 'get /foo' },
+            { http: 'gets /foo' },
+            {
+              http: {
+                path: '/home',
+                method: 'get',
+                invalidProp: 'baz',
+              },
+            },
+            {
+              http: {
+                method: 'get',
+              },
+            },
+            {
+              http: {
+                method: 'gets',
+                path: '/home',
+              },
+            },
+            {
+              http: null,
+              path: '/home',
+              method: 'get',
+            },
+          ],
+        },
+      },
+    };
+
+    const expectedErrorMessages = [
+      "Unsupported function event 'httpp' at functions.someFunc.events[0]",
+      'functions.someFunc.events[1].http should match pattern "^(get|post|put) [a-zA-Z0-9]+$"',
+      "Unrecognized property 'invalidProp' on 'functions.someFunc.events[2].http'",
+      "functions.someFunc.events[3].http should have required property 'path'",
+
+      // TODO: add list of allowed values as AJV error contains this list in error object
+      'functions.someFunc.events[4].http.method should be equal to one of the allowed values',
+
+      'Event should contain only one root property, but got 3 (http, path, method) at functions.someFunc.events[5].http',
```

Comment:
I think it's good enough to whoe error message as it comes directly from AJV, so:

```
Unrecognized property 'path', on `functions.someFunc.events[5]`
```

(I would not go with customization too far as it'll be hard to maintain)

Reply:
I customized it because you mentioned it in https://github.com/serverless/serverless/pull/7335/#pullrequestreview-407503581. If we remove customization then the message will be like this:
```
functions.someFunc.events[2].http should match schema any of: functions.someFunc.events[2].http should be string; functions.someFunc.events[2].http should be object
```

I think it is quite relevant. 

In light of that, please confirm you would like to remove this customization

Reply:
I think _"should match schema any of"_  is other AJV error, which in this case we should recognize, but we should rather expose deeper _"Unregnized property on"_  error here instead.

Isn't there also error simply informing about Unregnized property on `http` ? (same type as we get for unrecognized root property)

Reply:
> I think _"should match schema any of"_ is other AJV error, which in this case we should recognize, but we should rather expose deeper _"Unregnized property on"_ error here instead.

`functions.someFunc.events[5]` data path is on higher level than `functions.someFunc.events[5].http`. Current implementation groups errors around the most deepest error

Reply:
> Isn't there also error simply informing about Unregnized property on http ? (same type as we get for unrecognized root property)

Yes, such error is produced by AJV, with data path equals to `functions.someFunc.events[5]`. Also, AJV produces an error with `functions.someFunc.events[5].http` data path and `anyOf` keyword. Without customization, the deepest error would be chosen for reporting to the user

Reply:
This is a tricky part. Taking a scenario where we support two types for event (_string_ and _object_), and someone uses _object_ type, and puts an unrecognized property,  is there a clean way that we can recognize that it's just this _Unsupported property_ error (from those related to this event configuration) that should surface to user?

Reply:
I think I get your point. It's actually already implemented. The implementation works like this:
-  If user enters _object_ type and there is some _object_ related error, then such error will be reported 
- If user enters _string_ type and there is some _string_ related error, then such error will be reported
- In this particular case user enters _null_ type. So it does not fit the _string_ or _object_ pattern.

Reply:
Indeed in further comments I got lost and started referring to different case, sorry for confusion.

Now settling on this case, we deal with situation when on object which should have just one property (which states event type) more than one property is defined.

I think we also may go without any message customization, and pick all unrecognized property errors (which I think surface from AJV ?)

So in that case it'll be:
-  Unrecognized property 'path', on `functions.someFunc.events[5]`
-  Unrecognized property 'method', on `functions.someFunc.events[5]`

In such case we may also ignore any eventual event configuration errors (e.g. in case where configuration for `http` is set as object but with some unrecognized property) I would skip that, as event type configuration is invalid so it makes no sense to dive into validation of event properties

Reply:
I'm not sure I fully understand. I think what you are proposing is to customize not the message but the way of grouping errors. The current way of grouping errors relies on the assumption that the deepest error is the most relevant one. What you are suggesting (if I correctly understand) is to introduce an exception to this rule. It may be implemented, but then we would nee to remove `anyOf` error related to deeper `functions.someFunc.events[5].http` path. 

Don't you like the original message? I think it gives user a hint on how to fix it. The easiest will be to remove customization and show it to user instead.

Reply:
Yes, I think in this specific case (errors in event type configuration), we should discard any deeper errors (do not communicate any event property configuration errors)

Otherwise  approach when we show the depest errors in case of `anyOf` I think is totally valid, and we should apply it to all cases generically.

Wouldn't that be best for UX?

Reply:
> Yes, I think in this specific case (errors in event type configuration), we should discard any deeper errors (do not communicate any event property configuration errors)

Is it something we should deal with in this PR? I don't see the benefit of changing existing implementation of error grouping.

> Otherwise approach when we show the depest errors in case of anyOf I think is totally valid, and we should apply it to all cases generically.

Currently, it is customized. Do you think the current customized message is not good enough? I could try to change the customized message to look like this:
```
Unrecognized property 'path', on functions.someFunc.events[5];  Unrecognized property 'method', on functions.someFunc.events[5]
```
(technically there would be one message, not two)

> Wouldn't that be best for UX?

I think the original message (I pasted it in https://github.com/serverless/serverless/pull/7335#discussion_r450947853 ) is good for UX. So, I'd remove customization.

Reply:
I think custom handling of errors that's result of invalid event type configuration (in whatever form, typo, extra properties etc) should be in scope of this PR, and I'd try to do it right from the start.

Let's also separate two concerns:
- Customization of `anyOf` errors should work as we agreed but it should not be fixed or mixed in any way with custom handling of event type configuration errors.
- Event type configuration errors, while they're communicated through `anyOf` errors, should probably have its own customization which takes precednse over `anyOf` handling (I think it'll be good to not show deep level properties errors if event type configuration is erroneous)

>  think the original message (I pasted it in #7335 (comment) ) is good for UX

This message refers to configuration of event properties, while here deal with error in configuration of event type (which is different object).

Reply:
So, what I need to do is simply adjust the custom message under the line of code we discuss to look like this:
```
Unrecognized properties 'path', 'method' on functions.someFunc.events[5]
```

Hoping I got what you meant.

Reply:
Assuming that AJV exposes errors as: `Unrecognize property 'path' ...` and `Unrecognize property 'method'..`.

I would ensure that just those errors are surfaced. So in this given case of erroneous event type configuration I wouldn't customize any message, I would just apply custom filtering on errors.

I believe we need a custom message when user introduces event name that doesn't exists (as `httpp`), but in this case we know that valid event name was used, we just don't like extra properties.

Reply:
> I would ensure that just those errors are surfaced. So in this given case of erroneous event type configuration I wouldn't customize any message, I would just apply custom filtering on errors.

Is it a deal-breaker if I implement it via custom messages instead of adding logic to grouping errors? The reason I don't want with done via groping is that it would need to produce two errors, for `path` and `method`, instead of one with the deepest data path. Doing this via custom messages seem easier because we just need to analyze if user config and AJV errors meet some criteria, then the custom error is constructed.

Reply:
> I don't want with done via groping is that it would need to produce two errors, for path and method

So there are no such errors produced by AJV, which we may just surface?

Reply:
Here is list of errors: https://gist.github.com/preshetin/d1138a7db897658e7fa63b24fec5269a

Those errors were produced by this config:
```
functions:
  someFunc:
    handler: handler.main
    events:
      - http:
        path: /home
        method: get
```

Reply:
I'll try to implement it via grouping.

Reply:
Ok, so one assumption I took was wrong. I thought that `Unrecognized property "x"...` directly comes from AJV.
While it's our message customization for `additionalProperties` error, right ? If so, that is great, we should definitely keep that.

Now, in light of that, my idea was to, in given case, surface those two `additionalProperties` errors.
Having that, we should naturally have their message customized into `Unrecognized property "x"...` (with same logic, through which we customize messages for all `additionalPropeties` errors we surface)

Reply:
> While it's our message customization for additionalProperties error, right ? If so, that is great, we should definitely keep that.

Yes, our code contains text like `Unrecognized property "x"..`. It adjusts it because initial AJV eroror does not have `'x'` in its default message while is useful to show. Same with `enum` types of errors: such errors are now adjusted to include list of allowed values which wasn't included in default AJV error.

___

I ended up implementing the initial issue via custom messages instead of modifying grouping implementation. I believe such approach is better because I treat grouping as more 'fundamental' rather than customizing messages which I treat more high level. 

If you still insist on reimplementig it with grouping... just let me know one more time. I'll do this your way then.